### ISSUE TITLE
Unexpected itemsize creating bytes_ array on Python 3

### ISSUE THREAD

**Issue by taldcroft** [numpy/numpy#8403]
In Python 3, creating an array from a `bytes_` object always increases the itemsize by 3, for example:
```python
In [24]: np.array([np.bytes_(b'ab')])
Out[24]: 
array([b'ab'], 
      dtype='|S5')
```
This does not happen in Python 2, nor when creating an array from an equivalent Python bytes object in Py3:
```python
In [25]: np.array([b'ab'])
Out[25]: 
array([b'ab'], 
      dtype='|S2')
```
Tested with Python 3.5.2 and numpy 1.11.2.

**Comment by charris** [comment]
Oh, man, that's strange
```
>>> np.array([np.bytes_(b'abc')]).itemsize
6
>>> np.array(np.bytes_(b'abc')).itemsize
3
>>> np.array(b'abc').itemsize
3
>>> np.array([b'abc']).itemsize
3
```

Probably related to this
```
>>> type([np.bytes_(b'abc')][0])
<class 'numpy.bytes_'>
>>> type([b'abc'][0])
<class 'bytes'>
```

**Comment by charris** [comment]
I think the best solution would be to send `bytes_` to the moon, and then nuke it. Probably not acceptable in the short term, however. My suspicion is that somehow we are seeing utf16 in bytes. What I can't figure out is why it exists, but it seems to be defined in numeric types and then not used hardly anywhere.

Also
```
In [16]: np.array([np.bytes_(b'123')], dtype='S')
Out[16]: 
array([b'123'], 
      dtype='|S3')
```

**Comment by charris** [comment]
And
```
class bytes_(builtins.bytes, character)
```
I suspect it is the `character` that is causing problems. Using `bytes` instead of `bytes_` works fine.

**Comment by seberg** [comment]
my suspicion is that we use `repr` to figure out the size somewhere, and the repr includes the `b''` stuff meaning 3 characters more. No idea where this really happens though.

**Comment by seberg** [comment]
No, wait that seems nonsense, then it should add two characers in python two I suppose.

**Comment by seberg** [comment]
ah:
```
# python3:
In [3]: str(np.bytes_(b'ab'))
Out[3]: "b'ab'"
```

**Comment by charris** [comment]
@seberg Nice catch, I bet you are right.

**Comment by sac111gp** [comment]
Hi, any updates here?
We observed this with 1.16 version of numpy too.

**Comment by eric-wieser** [comment]
Looks like this fell though the cracks, thanks for the ping.

**Comment by sethtroisi** [comment]
I spent sometime investigating this with breakpoints and debug statements littered all over but I wasn't able to find it. These were my debug statements

https://github.com/numpy/numpy/compare/master...sethtroisi:len_bytes

I was using this `test.py` to debug
```
import numpy as np
  


for i in range(100):
    print()

z = np.bytes_(b'ab')
print (f"z = '{z}' | dtype: {z.dtype}, len(z) = {len(z)}")
a = np.array([z])
print (f"np.array([z]) = '{a}'")
print (f"a[0] = '{a[0]}'")
print (f"list(a[0]) = '{list(a[0])}' chr(97) = {chr(97)}")

# This code makes sense because
#   str(b'foobar') => "b'foobar'"
'''
tests = ["a", "foo", "foobar"]
tests = ["foobar"]

for a in tests:
#    byt = np.bytes_(a)
    byt = np.string_(a)
    s_byt = str(byt)
    import pdb; pdb.set_trace()

    lb = len(byt)
    ls = len(s_byt)
    print("lens: ", lb, ls)

    print (f"byt: '{byt}', str(byt): '{s_byt}'")
    print (f"type(byt): {type(byt)}, type(s_byt): {type(s_byt)}")
    print (f"len(str(np.bytes_({a}))):", len(s_byt))
    print (f"len(np.bytes_({a}))     :", len(byt))
```
then running it with `gdb --args python3 test.py` set a breakpoint somewhere and continue ~1100 times (for numpy importing) then stepping around. I remember being suspicious in `PyArray_DTypeFromObjectHelper` which has some complicated logic to determine if an object is an array or str or unicode...

**Comment by seberg** [comment]
@sethtroisi this is in C-code, and now it is in `string_discover_descr_from_pyobject` which calls `PyArray_DTypeFromObjectStringDiscovery`, which we did not update (although we did notice it not too long ago). Those are in `common.c` and `dtypemeta.c` currently.

**Comment by sethtroisi** [comment]
@seberg was this fixed along the way?
the original code sample is no longer reproducing in a clean branch for me.

```
>>> import numpy as np
>>> np.__version__
'1.20.0.dev0+593ef5f'
>>> np.array([np.bytes_(b'ab')])
array([b'ab'], dtype='|S2')
```

**Comment by seberg** [comment]
Yes, this was fixed as Seth noted, not sure whether in 1.20 or maybe even 1.19.  All of the examples above work fine now.
